src/cunumeric/sort/sort.cu

```diff
@@ -95,338 +137,1460 @@ struct SortPiece {
-                                 cudaMemcpyDeviceToDevice,
-                                 stream));
-      pos += size_send[r];
+    {
+      Buffer<VAL*> val_send_buffers_ptr = create_buffer<VAL*>(num_sort_ranks, Memory::Z_COPY_MEM);
+      for (size_t r = 0; r < num_sort_ranks; r++) {
+        val_send_buffers_ptr[r] = val_send_buffers[r].ptr(0);
+      }
+
+      // is either 1, 2, 4, ... , THREADS_PER_BLOCK/2
+      int segments_per_threadblock =
+        compute_segments_per_threadblock(segment_size_l, num_segments_l);
+      dim3 block_shape =
+        dim3(THREADS_PER_BLOCK / segments_per_threadblock, segments_per_threadblock);
+      const size_t num_blocks_y = (num_segments_l + block_shape.y - 1) / block_shape.y;
+      const size_t num_blocks_x = ((segment_size_l + 31) / 32 + block_shape.x - 1) / block_shape.x;
+      dim3 grid_shape           = dim3(num_blocks_x, num_blocks_y);
+      copy_data_to_merge_buffers<<<grid_shape, block_shape, 0, stream>>>(segment_blocks,
+                                                                         size_send,
+                                                                         local_sorted.values,
+                                                                         val_send_buffers_ptr,
+                                                                         num_segments_l,
+                                                                         num_segments_l_aligned,
+                                                                         segment_size_l,
+                                                                         my_rank,
+                                                                         num_sort_ranks);
+
+      if (argsort) {
+        Buffer<int64_t*> idc_send_buffers_ptr =
+          create_buffer<int64_t*>(num_sort_ranks, Memory::Z_COPY_MEM);
+        for (size_t r = 0; r < num_sort_ranks; r++) {
+          idc_send_buffers_ptr[r] = idc_send_buffers[r].ptr(0);
+        }
+        // need to sync as we share values in between host/device
+        copy_data_to_merge_buffers<<<grid_shape, block_shape, 0, stream>>>(segment_blocks,
+                                                                           size_send,
+                                                                           local_sorted.indices,
+                                                                           idc_send_buffers_ptr,
+                                                                           num_segments_l,
+                                                                           num_segments_l_aligned,
+                                                                           segment_size_l,
+                                                                           my_rank,
+                                                                           num_sort_ranks);
+        CHECK_CUDA(cudaStreamSynchronize(stream));  // needed before Z-copy destroy()?
```

Comment:
I'd make sure if this synchronization is really necessary. (same applies to other synchronization calls too)

Reply:
I removed some of the synchronization calls that were outdated. The 2 synchronizations related to Z-copy (especially the one in question here) are mandatory - the tests fail right away when I remove them.